File: 103.binary-tree-zigzag-level-order-traversal.py
#
# @lc app=leetcode id=103 lang=python3
#
# [103] Binary Tree Zigzag Level Order Traversal
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution: #这个是层序的代码，改吧改吧就好了
    def zigzagLevelOrder(self, root: TreeNode) -> List[List[int]]:
        if not root:
            return []
        # if not root.left and not root.left
        res = []
        current_level_values = []
        current_queue = []
        next_queue = []
        current_queue.append(root)

        index = 0
        while current_queue:
            node = current_queue.pop(index)
            if node.left:
                next_queue.append(node.left)
            if node.right:
                next_queue.append(node.right)
            current_level_values.append(node.val)
            if not current_queue:
                index = 0 if index == -1 else -1

                res.append(current_level_values)
                current_level_values = []
                if next_queue:
                    current_queue = next_queue
                    next_queue = []
        
        return res      
# @lc code=end

File: golang/103.binary-tree-zigzag-level-order-traversal.go
/*
 * @lc app=leetcode id=103 lang=golang
 *
 * [103] Binary Tree Zigzag Level Order Traversal
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func zigzagLevelOrder(root *TreeNode) [][]int {
	if root == nil {
		return [][]int{}
	}

	layer := 0
	curLayerNode := []*TreeNode{root}
	result := [][]int{}
	for len(curLayerNode) != 0 {
		curLayerValue := []int{}
		nextLayerNode := []*TreeNode{}
		{ // for zigzag traveersal

			// zig zag means in the opposite direction everytime 
			len := len(curLayerNode)
			//  exchange nodes[i] and nodes[n-1-i] in place
			for i := 0; i < len/2; i++ {
				tmp := curLayerNode[i]
				curLayerNode[i] = curLayerNode[len-1-i]
				curLayerNode[len-1-i] = tmp
			}
		}
		for _, node := range curLayerNode {
			curLayerValue = append(curLayerValue, node.Val)
			if layer%2 == 0 {
				if node.Left != nil {
					nextLayerNode = append(nextLayerNode, node.Left)
				}
				if node.Right != nil {
					nextLayerNode = append(nextLayerNode, node.Right)
				}
			} else {
				if node.Right != nil {
					nextLayerNode = append(nextLayerNode, node.Right)
				}
				if node.Left != nil {
					nextLayerNode = append(nextLayerNode, node.Left)
				}

			}
		}
		result = append(result, curLayerValue)
		curLayerNode = nextLayerNode
		layer = layer + 1
	}
	return result
}

// @lc code=end

